### PR TITLE
feat(chat-models): Tool calling feature for Ollama (#288)

### LINKED ISSUE #288: Add support for Ollama Tools
https://js.langchain.com/docs/integrations/chat/ollama_functions

> LangChain offers an experimental wrapper around open source models run locally via [Ollama](https://github.com/jmorganca/ollama) that gives it the same API as OpenAI Functions.

### PR DESCRIPTION
**Pull request by Ganeshsivakumar**
Resolves #288

### CODE REVIEW

**Review by davidmigloz — COMMENTED** [review]
Overall, it was great 🙌 My comments are mainly improvements. I'll push a new commit resolving them.

File: packages/langchain_ollama/lib/src/chat_models/ollama_tools.dart [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+///    'unit': {
+///      'type': 'string',
+///      'enum': ['celsius', 'fahrenheit'],
+///    },
+///  },
+///  'required': ['location'],
+///},
+///);
+///```
+class OllamaTools extends BaseChatModel<ChatOllamaToolOptions> {
```
[the remark this answers is not in the document]
Comment (davidmigloz) [reply packages/langchain_ollama/lib/src/chat_models/ollama_tools.dart]: fixed

File: packages/langchain_ollama/lib/src/chat_models/types.dart [hunk 2]
```diff
@@ -271,3 +274,62 @@ class ChatOllamaOptions extends ChatModelOptions {
     );
   }
 }
+
+/// {@template chat_ollama_tool_options}
+/// Options to pass into OllamaTools.
+/// {@endtemplate}
+class ChatOllamaToolOptions extends ChatModelOptions {
```
[the remark this answers is not in the document]
Comment (davidmigloz) [reply packages/langchain_ollama/lib/src/chat_models/types.dart]: I've split it into:
- `chat_ollama`
- `chat_ollama_tools`

File: packages/langchain_ollama/lib/src/chat_models/ollama_tools.dart [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+    final List<ChatMessage> messages, {
+    final bool stream = false,
+    final ChatOllamaOptions? options,
+  }) {
+    return GenerateChatCompletionRequest(
+      model: options?.model ??
+          defaultOptions.options.model ??
+          throwNullModelError(),
+      messages: messages.toMessages(),
+      format: options?.format?.toResponseFormat(),
```
[the remark this answers is not in the document]
Comment (davidmigloz) [reply packages/langchain_ollama/lib/src/chat_models/ollama_tools.dart]: done

File: packages/langchain_ollama/test/chat_models/ollama_tools_test.dart [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+          } catch (e) {
+            expect(
+              e.toString(),
+              'Exception: Model did not respond in vaild json string format, try improving your prompt, instruct to "respond in JSON"',
+            );
+          }
+        },
+      );
+      test(
+          'should return default tool if no other tools should be called for a given query',
```
[the remark this answers is not in the document]
Comment (davidmigloz) [reply packages/langchain_ollama/test/chat_models/ollama_tools_test.dart]: Now the `_conversational_response` tool is never returned. If the model calls it, the input of the tool is returned as the content of the `AIChatMessage`

File: packages/langchain_ollama/lib/src/chat_models/types.dart [hunk 5]
```diff
@@ -271,3 +274,62 @@ class ChatOllamaOptions extends ChatModelOptions {
...
+  ///[ChatOllamaOptions] that needs to be provided
+  final ChatOllamaOptions options;
+
+  ///It will be used as system message for model to instruct it on tool calling,
+  ///its optional to provide, a [defaultToolSystemPromtTemplate] will be used if its not provided
+  final String? toolSystemPromptTemplate;
+
+  ///Default system template that will be used
+  static const String defaultToolSystemPromtTemplate =
+      'You have access to the following tools: {tools} You must always select one of the above tools based on question and respond with only a JSON object matching the following schema:{{"tool": <name of the selected tool>,"tool_input": <parameters for the selected tool, matching the tools JSON schema>}}. Respond using JSON';
```
[the remark this answers is not in the document]
Comment (davidmigloz) [reply packages/langchain_ollama/lib/src/chat_models/types.dart]: Based on the `toolChoice` passed, it now customizes the tools specified in the prompt

**Review by davidmigloz — APPROVED** [review]

**Comment by davidmigloz** [comment]
Could you take a look at my changes?

Pending things:
- Port [the documentation](https://js.langchain.com/v0.2/docs/integrations/chat/ollama_functions)
- Streaming support (we can leave it outside the scope of this PR)

**Comment by Ganeshsivakumar** [comment]
Refactored changes for formatting prompt logic with tool choice option , parsing result, handling multiple tool result and updated folder structure looks good.

**Comment by davidmigloz** [comment]
I've added the documentation. It's good to go 🚀

**Pull request merged**